### DISCUSSION TITLE
NextAuth seperate backend

### DISCUSSION THREAD

**Discussion by georgekrax** [nextauthjs/next-auth#2971] [Help] [answered]
Hello everyone,

I have an application that uses NextAuth.js with the help of `/pages/api` server pages. However, the project is getting massive and has big scale, and I would like to split the frontend with the backend. Any ideas, on how this can work while keeping NextAuth. Should j get rid of it and implement something custom on my own?

Thanks in advance

**Comment by balazsorban44** [comment, marked as answer]
What do you mean exactly?

**Reply by georgekrax** [reply]
I have being using as backend the server provided by Next.js, that exists on `/pages/api` directory. Now, I want to take this backend and host it elsewhere, on a new project directory, because the project has scaled up a lot and I need to divide the code.

**Reply by balazsorban44** [reply]
What requirements do you have? 
Will it be on the same domain? 
Will you still use Next.js?

**Reply by praveshsapkota** [reply]
I have had a similar problem i had a existing graphql server with prisma database and i authenticated user in frontend using /api serverlessly and i need to authenticate users for there permission if there are authorized  to access api or not  and my solution i will state below its not perfect but it works 

You can see in session callback i am sending user role and a signed JWT which containes user information and its signed so that no one can temper it . you can access the session inside your react using useSession hook

you can send that signed JWT to server with http request header authorization : "bearer Token"

![nextauth](https://user-images.githubusercontent.com/24776096/137596696-7b55d7d0-a4ba-4798-a049-715995a4a6fe.png)


> Hello everyone,
> 
> I have an application that uses NextAuth.js with the help of `/pages/api` server pages. However, the project is getting massive and has big scale, and I would like to split the frontend with the backend. Any ideas, on how this can work while keeping NextAuth. Should j get rid of it and implement something custom on my own?
> 
> Thanks in advance

**Reply by georgekrax** [reply]
I will keep using Next.js as my frontend only. The backend will be a separate Node.js & Express API server. Yes, it will be on the same domain. The fronted will be `www.domain.com` and the backend (API) on `www.domain.com/graphql`.

**Reply by georgekrax** [reply]
@praveshsapkota Thank you for your helpful answer! I will explore it, and let you know if it works.

**Reply by praveshsapkota** [reply]
@georgekrax 
 which graphql client are you using in frontend

**Reply by georgekrax** [reply]
@praveshsapkota  Apollo. The issue seems to be with the `session` callback, as it is called every time to know whether the user is authenticated or not. So, I cannot make a request to the server there because it will cost me on performance.

**Reply by praveshsapkota** [reply]
as jwt contains same information it wont matter i have saved it in my localstorage you can save in cookie or something else 

and you can access its from inside your apollo client and set it as header when apollo makes request

setting token to localstorage 
![insideIndex](https://user-images.githubusercontent.com/24776096/137597030-fbce2e79-dc21-4bb4-85bc-dc8af44a7f00.png)

accessing token in apolloclient
![11111](https://user-images.githubusercontent.com/24776096/137597120-d1731e38-0d02-45c3-820b-0c5b8db44ae0.png)

**Reply by praveshsapkota** [reply]
> @praveshsapkota Apollo. The issue seems to be with the `session` callback, as it is called every time to know whether the user is authenticated or not. So, I cannot make a request to the server there because it will cost me on performance.

@georgekrax 
you dont need to make server call every time session changes you are just saving it in localstorage and when apollo makes request you can access it from localstorage and pass it as header

**Reply by georgekrax** [reply]
@praveshsapkota  Yes, but I would not like to use `localStorage` for safety reasons. Can I implement the same with cookies?

**Reply by praveshsapkota** [reply]
i have similar project as your i guess you can get some ideas its not perfect as i am learning but its kind of working 😄
repo link frontend : - https://github.com/praveshsapkota/ktm-sale-frontend
backend link : - https://github.com/praveshsapkota/KtmSale-Backend

**Reply by praveshsapkota** [reply]
> @praveshsapkota Yes, but I would not like to use `localStorage` for safety reasons. Can I implement the same with cookies?

yes you can you just need to add cookie.js and use Cookie.set("jwt",your Token in use effect )

**Reply by georgekrax** [reply]
@praveshsapkota  Okay, thank you very much for you help!

**Reply by praveshsapkota** [reply]
glad to help  😀

@georgekrax 
i guess you can make this discussion  answered  and change the discussion topic to something more describing so it might help others too

**Reply by balazsorban44** [reply]
> I have had a similar problem i had a existing graphql server with prisma database and i authenticated user in frontend using /api serverlessly and i need to authenticate users for there permission if there are authorized  to access api or not  and my solution i will state below its not perfect but it works 
> 
> You can see in session callback i am sending user role and a signed JWT which containes user information and its signed so that no one can temper it . you can access the session inside your react using useSession hook
> 
> you can send that signed JWT to server with http request header authorization : "bearer Token"
> 
> ![nextauth](https://user-images.githubusercontent.com/24776096/137596696-7b55d7d0-a4ba-4798-a049-715995a4a6fe.png)
> 
> 
> > Hello everyone,
> > 
> > I have an application that uses NextAuth.js with the help of `/pages/api` server pages. However, the project is getting massive and has big scale, and I would like to split the frontend with the backend. Any ideas, on how this can work while keeping NextAuth. Should j get rid of it and implement something custom on my own?
> > 
> > Thanks in advance
> 
> 

If you decide to run your own auth system, hopefully you don't save passwords in plain text as you do on the image. 👀 I recommend against using the credentials flow unless you have a dedicated person/team concerning security. storing passwords correctly is crucial.

You also seem to leak the `userAccount` between calls. I don't have much experience, but I feel you shouldn't be doing that either. You could refactor the code to be able to save that only in the context of the invocation of the endpoint.

same with `apiToken`

Furthermore, dotenv isn't necessary in Next.js

**Reply by praveshsapkota** [reply]
@balazsorban44 

I havent completely implemented credentials provider it was there just for the custom login page 

about passwords i will hash and save in db and would compare its when signing in 

![authorize](https://user-images.githubusercontent.com/24776096/137599004-0b0e0fbc-dde7-4d5b-8b8a-2781a5bc1ca5.png)


and about leaking userAccount and apiToken 

**i didnt quite get about saving only in context of invocation of endpoint** 
can you please elaborate or point me to some sort of docs 

thanks for the suggestions 😅

**Reply by balazsorban44** [reply]
```js
export default async function auth(req, res) {
  // save anything here
  return await NextAuth(req, res, {
    //... your options
  })
}
```

**Reply by praveshsapkota** [reply]
👍🏻

**Comment by georgekrax** [comment]
@balazsorban44 I also have another question. Can I create a custom OAuth provider and pass it to NextAuth.js, as I have a second backend which is where "true" really authentication happens?

**Reply by balazsorban44** [reply]
https://next-auth.js.org/configuration/providers/oauth-provider#using-a-custom-provider

https://next-auth.js.org/configuration/providers/oauth-provider#options

**Reply by georgekrax** [reply]
Oh ok, thank you very much! But how should the OAuth server return in order for this to work? It is a GraphQL server, should I implement something differently?

> Note: The custom backend OAuth server does not have a front-end like Google or Facebook, there is just the GraphQL mutation resolver, where also the password and username are passed as arguments

**Reply by balazsorban44** [reply]
I don't know if you can implement OAuth on top of GraphQL, or if it's worth the effort. If you don't know what you are doing, I would rather recommend finding a tested library, instead of creating your own.

But I also have a feeling that you actually meant to use your GraphQL endpoint to simply receive API requests that might need authorization. For that, you can attach an auth token in the requests' `Authorization` header, and following your library's guide, you can validate that.

If you actually meant to authenticate with GraphQL, you probably don't talk about OAuth but simple credentials flow, which we also advocate against, due to the extra security concerns it brings with itself, without a dedicated person working on that. In this case, You should have a look at the Credentials provider instead: https://next-auth.js.org/providers/credentials

**Reply by georgekrax** [reply]
Okay. Thank you for your response. I think that I will implement something with the Credentials provider, and then when the OAuth server has a frontend, I will add a custom provider.

**Reply by georgekrax** [reply]
Would you mind to tell me when the `authorize()` function of the Credentials provider is invoked. I will make a DB request there, and I do not want it to run many times (for performance reasons).

**Reply by balazsorban44** [reply]
it's only run when the user is being authorized, so once at login. after that, only the jwt and session callbacks will be invoked

**Reply by georgekrax** [reply]
Oh, okay thank you so much for your help!

**Reply by georgekrax** [reply]
One last question, is it possible to return a JWT token in the `authorize()` function, which I have created, and do not pass a `jwt()` callback?

**Reply by balazsorban44** [reply]
Whatever you return in authorize will be available in the jwt callback, but you can save your own jwt in there without a problem, as long as you remain under 4096 bytes which is the maximum cookie size.

**Reply by georgekrax** [reply]
Okay, so I can directly return my token in the `jwt()` callback. Can NextAuth also handle refresh tokens?

**Reply by balazsorban44** [reply]
Not currently https://next-auth.js.org/tutorials/refresh-token-rotation

a note if you look at the above link to implement it yourself, I suggest encrypting the generated JWT as it should not be exposed to the frontend in any way.

**Reply by georgekrax** [reply]
Oh okay. Thank you for all of your help!
